File: docs/review/period.js
/** Payday periods. See CAPTURE_AND_SCREENS.md. */

export function clampDay(year, month, day) {
  const last = new Date(year, month, 0).getDate();
  return Math.min(day, last);
}

export function parseIso(iso) {
  const [y, m, d] = iso.split("-").map(Number);
  return { y, m, d };
}

export function toIso(y, m, d) {
  return `${y}-${String(m).padStart(2, "0")}-${String(d).padStart(2, "0")}`;
}

export function addMonths(y, m, delta) {
  const date = new Date(y, m - 1 + delta, 1);
  return { y: date.getFullYear(), m: date.getMonth() + 1 };
}

export function dayBefore(iso) {
  const { y, m, d } = parseIso(iso);
  const date = new Date(y, m - 1, d - 1);
  return toIso(date.getFullYear(), date.getMonth() + 1, date.getDate());
}

export function periodContaining(iso, paydayMode, paydayDay) {
  const { y, m, d } = parseIso(iso);
  if (paydayMode !== "day" || !paydayDay) {
    const start = toIso(y, m, 1);
    const next = addMonths(y, m, 1);
    const end = dayBefore(toIso(next.y, next.m, 1));
    return { start, end };
  }
  const clamped = clampDay(y, m, paydayDay);
  let startY = y;
  let startM = m;
  if (d < clamped) {
    const prev = addMonths(y, m, -1);
    startY = prev.y;
    startM = prev.m;
  }
  const startDay = clampDay(startY, startM, paydayDay);
  const start = toIso(startY, startM, startDay);
  const next = addMonths(startY, startM, 1);
  const nextStart = toIso(next.y, next.m, clampDay(next.y, next.m, paydayDay));
  return { start, end: dayBefore(nextStart) };
}

export function inPeriod(iso, period) {
  return iso >= period.start && iso <= period.end;
}

const MONTHS = ["Jan", "Feb", "Mar", "Apr", "May", "Jun", "Jul", "Aug", "Sep", "Oct", "Nov", "Dec"];

export function formatDay(iso) {
  const { y, m, d } = parseIso(iso);
  return `${d} ${MONTHS[m - 1]} ${y}`;
}

export function formatDayShort(iso) {
  const { m, d } = parseIso(iso);
  return `${d} ${MONTHS[m - 1]}`;
}

export function formatPeriod(period) {
  const a = parseIso(period.start);
  const b = parseIso(period.end);
  if (a.y === b.y) return `${a.d} ${MONTHS[a.m - 1]} – ${b.d} ${MONTHS[b.m - 1]}`;
  return `${formatDayShort(period.start)} ${a.y} – ${formatDayShort(period.end)} ${b.y}`;
}

export function todayIso() {
  const now = new Date();
  return toIso(now.getFullYear(), now.getMonth() + 1, now.getDate());
}

export function addDays(iso, days) {
  const { y, m, d } = parseIso(iso);
  const date = new Date(y, m - 1, d + days);
  return toIso(date.getFullYear(), date.getMonth() + 1, date.getDate());
}

File: lib/core/time/period.dart
class LocalDate {
  const LocalDate(this.year, this.month, this.day);

  final int year;
  final int month;
  final int day;

  factory LocalDate.parse(String iso) {
    final parts = iso.split('-');
    return LocalDate(int.parse(parts[0]), int.parse(parts[1]), int.parse(parts[2]));
  }

  factory LocalDate.today() {
    final now = DateTime.now();
    return LocalDate(now.year, now.month, now.day);
  }

  String get iso =>
      '$year-${month.toString().padLeft(2, '0')}-${day.toString().padLeft(2, '0')}';

  DateTime get asDate => DateTime(year, month, day);

  LocalDate addDays(int days) {
    final next = asDate.add(Duration(days: days));
    return LocalDate(next.year, next.month, next.day);
  }

  LocalDate addMonths(int delta) => LocalDate(year, month + delta, 1);

  int get lengthOfMonth => DateTime(year, month + 1, 0).day;

  int clampDay(int payday) => payday > lengthOfMonth ? lengthOfMonth : payday;

  bool operator >=(LocalDate other) => iso.compareTo(other.iso) >= 0;
  bool operator <=(LocalDate other) => iso.compareTo(other.iso) <= 0;
}

class Period {
  const Period(this.start, this.end);
  final LocalDate start;
  final LocalDate end;
  bool contains(LocalDate date) => date >= start && date <= end;
}

Period periodContaining(LocalDate date, {required String paydayMode, int? paydayDay}) {
  if (paydayMode != 'day' || paydayDay == null) {
    final start = LocalDate(date.year, date.month, 1);
    final next = LocalDate(date.year, date.month + 1, 1);
    return Period(start, next.addDays(-1));
  }
  var startYear = date.year;
  var startMonth = date.month;
  if (date.day < LocalDate(date.year, date.month, 1).letDay(paydayDay)) {
    final prev = DateTime(date.year, date.month - 1, 1);
    startYear = prev.year;
    startMonth = prev.month;
  }
  final startDay = LocalDate(startYear, startMonth, 1).letDay(paydayDay);
  final start = LocalDate(startYear, startMonth, startDay);
  final nextMonth = DateTime(startYear, startMonth + 1, 1);
  final nextStartDay = LocalDate(nextMonth.year, nextMonth.month, 1).letDay(paydayDay);
  final nextStart = LocalDate(nextMonth.year, nextMonth.month, nextStartDay);
  return Period(start, nextStart.addDays(-1));
}

extension on LocalDate {
  int letDay(int payday) => clampDay(payday);
}
